### DISCUSSION TITLE
Client binding: propose new describe() API

### DISCUSSION THREAD

**Discussion by fantix** [geldata/gel#4440] [Pre-RFC]
I'm proposing to add a new client-binding API `describe()`, which takes a query string, and returns the type information of the input/output to/from the query.

## Motivation

This API will power the code generator, and could potentially be used by the user's code.

## Design

The API should return 4 parts, the input typing information, the output typing information, the output cardinality and the capabilities of the query.

Starting from he output cardinality, it can be one of the five values below:

* `NO_RESULT`
* `AT_MOST_ONE`
* `MANY`
* `ONE`
* `AT_LEAST_ONE`

For input/output typing information, they should share the same recursive data structure that contains the following information:

* Type of the descriptor, one of the type descriptors as defined in [the docs](https://www.edgedb.com/docs/reference/protocol/typedesc), indicating the "type" of this type, e.g. if it is an object, or a named tuple.
* The Descriptor ID of the type as a UUID value.
* Optionally, the full schema name of the type. If module aliases is in use, the server will translate it into the real module names. The implementation is for now limited to using the `INJECT_OUTPUT_TYPE_NAMES` flag in the `Parse` message, and the server will return type names for 1) non-builtin base scalar types 2) all enum types. The implementation could also translate known type IDs of builtin base scalar types into their full schema names, like `std::int32`.

Depending on the type of the descriptor, the type information may include:
* The full schema name of the base type from which the current type is derived from.
* All the label strings if the current type is an enum type.
* Elements of the current container type, like objects, tuples, arrays or sets. Each element has recursively the same information as listed here, and also has the following in addition:
  * Name of this element, if the type is object or named tuple.
  * Cardinality for the value of the element on the object type. The cardinality value is the same as the output cardinality.
  * A boolean flag indicating if this element is an implicit field in the containing object type, when the [`CompilationFlag` is set](https://www.edgedb.com/docs/reference/protocol/messages#ref-protocol-msg-parse).
  * The kind of the pointer in the containing object type, can be one of `LINK`, `PROPERTY` or `LINK_PROPERTY`.

## Example API

For example, the Python client defines the following:

```python
class Cardinality(enum.Enum):
    NO_RESULT = 0x6e
    AT_MOST_ONE = 0x6f
    ONE = 0x41
    MANY = 0x6d
    AT_LEAST_ONE = 0x4d

class AnyType:
    desc_id: uuid.UUID
    name: typing.Optional[str]


class ElementKind(enum.Enum):
    LINK            = 1
    PROPERTY        = 2
    LINK_PROPERTY   = 3

class Element:
    type: AnyType
    cardinality: enums.Cardinality
    is_implicit: bool
    kind: enums.ElementKind

class ObjectType(AnyType):
    elements: typing.Dict[str, Element]

class SparseObjectType(ObjectType):
    pass


class SequenceType(AnyType):
    element_type: AnyType

class SetType(SequenceType):
    pass

class ArrayType(SequenceType):
    pass


class BaseScalarType(AnyType):
    pass

class ScalarType(AnyType):
    base_type: BaseScalarType


class TupleType(AnyType):
    element_types: typing.Tuple[AnyType]

class NamedTupleType(AnyType):
    element_types: typing.Dict[str, AnyType]


class EnumType(AnyType):
    members: typing.Tuple[str]


class RangeType(AnyType):
    value_type: AnyType


class DescribeResult(typing.NamedTuple):
    input_type: typing.Optional[AnyType]
    output_type: typing.Optional[AnyType]
    output_cardinality: enums.Cardinality
    capabilities: enums.Capability

class Client(...):
    ...
    def describe_query(
        self, query: str, *, inject_type_names: bool=False
    ) -> DescribeResult:
        ...
```

**Comment by i0bs** [comment]
Honestly a huge fan of this proposal and I'm absolutely for it, however, I have a very random question: since the minimum supported Python version is 3.10 for the repo, is there a reason why `Optional[T]` and `Tuple[T]` are used instead of `T | None`/`tuple[T]`, or is it preference?

**Reply by fantix** [reply]
Oh this will be [a part](https://github.com/edgedb/edgedb-python/pull/363) of edgedb-python which still supports back to Python 3.7.

**Comment by 1st1** [comment]
> Optionally, the full schema name of the type. The implementation is recommended to set the INJECT_OUTPUT_TYPE_NAMES flag in the Parse message, and let the server return type names for 1) non-builtin base scalar types 2) all enum types. The implementation could also translate known type IDs of builtin base scalar types into their full schema names, like std::int32.

I think we should add a keyword-only argument to `client.describe()` to inject names, as in `client.describe(query, inject_type_names=True)`

**Reply by 1st1** [reply]
> The implementation could also translate known type IDs of builtin base scalar types into their full schema names, like std::int32.

Sure, makes sense.

**Reply by tailhook** [reply]
What is the reason of making typenames in describe optional? It's not performance, right? As much as a see, it's just makes types more confusing (i.e. `AnyType.name` is optional depending on argument of a function).

**Reply by tailhook** [reply]
Also as far as I remember setting `INJECT_OUTPUT_TYPE_NAMES` will not return type names in `Parse`. It injects type names into the query result (and that make sense to find out actual type name of inherited object). But we still don't have type names in descriptors, do we?

**Reply by 1st1** [reply]
> What is the reason of making typenames in describe optional?

Because enabling that changes the shape of your query (type names get injected). So you're essentially introspecting a slightly different query with them.

> Also as far as I remember setting INJECT_OUTPUT_TYPE_NAMES will not return type names in Parse. It injects type names into the query result (and that make sense to find out actual type name of inherited object).

Correct.

> But we still don't have type names in descriptors, do we?

We have names in descriptors for some scalar types (such as enum).

**Comment by 1st1** [comment]
@fantix I know that I myself advocated for exposing the codec tree directly, but now I think we'd be exposing too much of an implementation detail.

How about we do this:

```python

class Type:
    id: uuid.UUID
    base_type_name: typing.Optional[str]
    sub_types: typing.List[SubType]
    codec: AbstractCodec
    name: typing.Optional[str]
    cardinality: typing.Optional[enums.Cardinality]
    is_implicit: bool
    pointer_kind: typing.Optional[enums.PointerKind]

class EnumType(Type):
    labels: typing.Tuple[str]
```

It will be just one quick tree traversal in the `decribe()` implementation which is totally fine, it's not a high-performance kind of method anyways.

**Reply by fantix** [reply]
Sure, that makes sense

**Reply by fantix** [reply]
Updated above in the proposal.

**Comment by tailhook** [comment]
Does it works in `query` mode only? I mean we have `query_single` and `query_required_single`. Do we expect it to be trivial to find out how result type is affected by that mode? Json modes always return the same type. But their input type can also be introspected. Again, do we expect people to run normal describe and ignore result types to find out info about json queries?

Why do we only have `inject_type_names` flag? Type ids and object ids are also optional? Is this specific to python bindings? (i.e. always use object id and never use type id?)

**Reply by tailhook** [reply]
Oh, well, is this `inject_type_names` not a query mode, but rather means `inject typenames into a type descriptor`?

(anyway my point about all those compilation modes still holds, Including `INJECT_OUTPUT_TYPE_NAMES` flag we have in the query compilation flags)

**Reply by fantix** [reply]
`INJECT_OUTPUT_TYPE_NAMES` will:
1. Add type annonations (of type names) in type descriptor for a) non-builtin base scalar types, and b) all enum types.
2. Add `__tname__` implicit field to objects.

(1) affects only the type descriptor, so it's only `CommandDataDescription`; (2) adds `__tname__` field to type descriptors (without the actual type of the object), and the actual type as the value of `__tname__` in `Data` messages.

> Does it works in `query` mode only? I mean we have `query_single` and `query_required_single`.

The proposed API shall do `Parse` only, and it should not assume any expected cardinality. So yes, the proposed API is equivalent to running the first half of `query()` without the `Execute` part (also without caching parsed result).

> Do we expect it to be trivial to find out how result type is affected by that mode?

Yeah I think so, the user could simply judge by `output_cardinality`:
* `ONE` -> `query_required_single`
* `AT_MOST_ONE` -> `query_single`
* `MANY` -> `query`
* `AT_LEAST_ONE` -> `query`
* `NO_RESULT` -> `execute`

> Json modes always return the same type. But their input type can also be introspected. Again, do we expect people to run normal describe and ignore result types to find out info about json queries?

In this case, theoretically we could (should?) return a JSON Schema for `output_type`, while the `input_type` is still the current `AnyType`, but I'd rather leave the JSON support to a new API like `describe_json()`.

**Reply by tailhook** [reply]
> `INJECT_OUTPUT_TYPE_NAMES` will:
> 
> 1. Add type annonations (of type names) in type descriptor for a) non-builtin base scalar types, and b) all enum types.
> 
> 2. Add `__tname__` implicit field to objects.
> 

So it will work if I want `INJECT_OUTPUT_TYPE_NAMES` for my query (can I even do that in current python bindings?)

But it will not work if I just want type names in the output of `.describe()`, because (2) will change the result type and it doesn't match want `.query()` returns any more.

**Reply by fantix** [reply]
> So it will work if I want `INJECT_OUTPUT_TYPE_NAMES` for my query

Yes.

>  can I even do that in current python bindings?

Negative, we turned that off internally.

> But it will not work if I just want type names in the output of `.describe()`, because (2) will change the result type and it doesn't match want .query() returns any more.

Well that's true, but we could filter out the "implicit" object shape elements (`id`, `__tname__`) before returning from `.describe()` API.

**Reply by tailhook** [reply]
> Negative, we turned that off internally.
>

So why would we have `inject_type_names` flag?
 
> > But it will not work if I just want type names in the output of `.describe()`, because (2) will change the result type and it doesn't match want .query() returns any more.
> 
> Well that's true, but we could filter out the "implicit" object shape elements (`id`, `__tname__`) before returning from `.describe()` API.

I still believe it's better to have `ANNOTATE_DESCRIPTORS` flag instead. Is it critical to have it working on current EdgeDB rather than 3.0?

**Reply by tailhook** [reply]
Also, this is a thread about different compilation modes.

I think we can have something like this `describe_query('<query text>', mode='query_required_single')`

**Reply by 1st1** [reply]
> So why would we have inject_type_names flag?

It also triggers annotating scalar types in the descriptor. It's a flag that triggers two things:

* inject `__tname__` for shapes into your results
* inject scalar type names to type descriptor

**Reply by fantix** [reply]
> Also, this is a thread about different compilation modes.
>
>I think we can have something like this `describe_query('<query text>', mode='query_required_single')`


`query_single()` only set `expected_cardinality` to `AT_MOST_ONE` in `Parse` and `Execute` messages, while `query_required_single()` does no more than that, except some additonal client-side checks. For now, on the server-side, `expected_cardinality=AT_MOST_ONE` only raises an error if the result is many, no other impact to the compilation at this moment. So I'd prefer not to add a `mode` or `expected_cardinality` parameter to the `describe_query()` API just yet.

`expected_cardinality=AT_MOST_ONE` does affect the result if `output_format=JSON`, but again I'd do that in a different `describe_query_json()` API.

Btw, `output_format=NONE` also affects the compilation, which is used by the `execute()` API.

**Comment by tailhook** [comment]
Should include `capabilities` in `DescribeResult`. This is quite useful information about the query.

**Reply by 1st1** [reply]
+1, @fantix

**Comment by tailhook** [comment]
Is `AnyType.id` a real type id or is it descriptor's id (the hash of all the fields in the shape and object type)? I think we don't have real type ids in current descriptors, right?

**Reply by fantix** [reply]
I'm sorry but I didn't get this one - why would the "real type id" be different from the type id in descriptors (the `id` field, not the `type` field in descriptors, per [doc](https://www.edgedb.com/docs/reference/protocol/typedesc) - the `type` field is simply represented by the Python class, like `EnumType`, `ObjectType` etc.)?

**Reply by tailhook** [reply]
Type descriptors currently contain only `Descriptor ID`: https://www.edgedb.com/docs/reference/protocol/typedesc 
While if you do `INJECT_IMPLICITY_TYPE_IDS` the data contains `__typeid__` matching `id` from the `schema::ObjectType`, and this is what I would probably need for codegen.

**Reply by fantix** [reply]
Jeez you're right 😲 I always thought the "Descriptor ID" is the real type ID 🤦‍♂️ but why are they different in the first place? They are both like a UUID5/SHA1 of a string concatenating the type name and subtypes, but done in slightly different ways, and base scalar types do share the same IDs.

> While if you do INJECT_IMPLICITY_TYPE_IDS the data contains __typeid__ matching id from the schema::ObjectType, and this is what I would probably need for codegen.

The thing is, codegen shouldn't really execute (`Execute`) the query (the only way to get real type IDs as `__tid__` field value), but only describing (`Parse`) won't get the "real type ids" (but the "Descriptor IDs"). 

Considering we don't really need to match the "Descriptor ID" to anything after codegen, I think it is okay we live with "Descriptor IDs", it's just that we should really document it correctly.

**Reply by tailhook** [reply]
Well, my idea was to codegen something like this:
```python
def get_all_fruits(conn):
   for item in conn.query('...'):
     if item.__typeid__ == '79bb8383-8098-4029-9af3-7a920a0ac57b':
         yield Apple(**item)
     # ...
```
But this does not work because object id's aren't stable between reincarnations of the same schema in another database as far as I understand.

Type ids can still be useful to find out which queries return same schema type and which are different. Although, maybe it's better to discourage their use and rely on type names for that task instead.

**Reply by tailhook** [reply]
I think renaming `AnyType.id` -> `AnyType.typedesc_id` or similar is good for the documentation.

**Comment by tailhook** [comment]
Also I think that for codegen task it should be possible to get the typenames of all types that can be in that collection. I mean: every set can contain objects of arbitrary number of types inherited from some single type.

Or do we expect users to run `DESCRIBE SOMETHING` query to get that info? Do we even have such a describe query? (i.e. fetch me all types inherited from this type)

**Reply by fantix** [reply]
We don't have the `DESCRIBE query` afaik, and it's not quite practical to "get typenames of all types that can be in that collection", please see also [this comment](https://github.com/edgedb/edgedb/pull/4325#issuecomment-1241313858).

**Reply by tailhook** [reply]
> We don't have the `DESCRIBE query` afaik

I meant something along the lines of `DESCRIBE MyBaseType GET ALL SUPERCLASSES`. But now I have to rethink it according to #4325 (which I'm not sure I'm on board with).

**Reply by 1st1** [reply]
> I meant something along the lines of DESCRIBE MyBaseType GET ALL SUPERCLASSES. But now I have to rethink it according to https://github.com/edgedb/edgedb/pull/4325 (which I'm not sure I'm on board with).

Hm, you can write a trivial introspection query for that.

**Comment by tailhook** [comment]
Also `.describe()` might be too similar to our `DESCRIBE OBJECT/WHATEVER` queries? So we should either support this pattern somehow i.e. `.describe_query()`, `.describe_object()`, or invent entirely new name for the method.

**Reply by fantix** [reply]
I think `describe_query()` makes sense.

**Reply by 1st1** [reply]
+1

**Comment by tailhook** [comment]
> An optional name of this sub type, this is usually the pointer name on object types or named tuples of different members.

Why it's not a property of the pointer itself? (`Element` in your schema)

I'm confused by this text. Is `AnyType.name` this name? Or is that the name of the type used `CREATE TYPE` (i.e schema name) and you've forgot `Element.name`?

**Reply by fantix** [reply]
Oh I'm sorry, this text is out of sync with the example - I used to have a `SubType` class with  a `.name` property. I'll edit this text.

**Comment by tailhook** [comment]
> Optionally, the full schema name of the type. 

Please make sure to document whether we translate `with_module_aliases` in this name. We do, right? It's done server-side, right?

**Reply by fantix** [reply]
Yes it's done on the server-side. I'll clarify like "if module aliases are used, they will be translated to the real module names".

**Comment by tailhook** [comment]
Why you use `PointerKind`, but `Element` and `elements`? 

Also why `Tuple` and `NameTuple` has `sub_types`? They are more like "elements" than "sub types", no? (i.e. i would likely confuse "sub type" term with "sub class"). Naming them `elements` is fine even if `ObjectType` would contain `pointers: List[Pointer]`.

**Reply by fantix** [reply]
> Why you use `PointerKind`, but `Element` and `elements`?

This is messed up in the progress - elements from type descriptors doc, and pointer from `edgedb/introspect.py` asl well as the server source code. I think we should just all go for "elements" as that is already advertised in the type descriptor doc.

> Also why Tuple and NameTuple has sub_types? They are more like "elements" than "sub types", no?

You're absolutely right, "elements" it is.

**Comment by 1st1** [comment]
@fantix @tailhook I think we should go with the proposed API (with most of the actionable comments in this thread integrated, obviously). We can add more fields to the introspection structs later, and maybe starting with 3.0 we can push more data in our type descriptors. But I don't want to wait until 3.0 to ship this.

I also want to release the new edgedb-python with codegen. Can we do that with `describe_query()` in the next couple of days, or should we just release it with `_describe_query()` - undcoumented and "private"?

**Reply by fantix** [reply]
Proposal at the top, and [the PR](https://github.com/edgedb/edgedb-python/pull/363) are both updated now.

> or should we just release it with `_describe_query()` - undcoumented and "private"?

No document added yet, so we can do either way.

**Reply by tailhook** [reply]
> We can add more fields

The problem with adding more fields later is that those fields will be optional in the type system. And users have to deal with that somehow, even if they will never use EdgeDB < 3.0. So while it's cheap for us, it isn't so much for users.

Having private API is fine.
